### DISCUSSION TITLE
Fastest way to count/access events in a midi file?

### DISCUSSION THREAD

**Discussion by jeroen** [FluidSynth/fluidsynth#1306] [General]
I wrote some bindings for the R programming language. In one function we read a midi file into an in-memory table object.

Currently this is implemented by playing the midi and using `fluid_player_set_playback_callback` to store events. We actually have to `fluid_player_play()` twice, because we first need to know how much memory to allocate to store the midi data. 

On the first run we set `fluid_player_set_playback_callback` to a function that only increments a counter for each event. Once complete we know the total number of events in the midi file. We then reset the player, allocate the output memory, and set the callback to another function that stores each event in our output table, and `fluid_player_play()` again.

The code in here: https://github.com/ropensci/fluidsynth/blob/main/src/readmidi.c

Some questions:

 - Is there a better way to count the number of events in a midi file, without having to play the file?
 - Is there a better way to read events in a midi file than playing it, if we are not interested in synthesising?
 - Is there a way to access metadata events in the midi file, which do not appear in `fluid_player_set_playback_callback` ?

**Comment by mawe42** [comment]
Do you have to use FluidSynth to read the MIDI data? Maybe you could simply use existing R libraries, like tuneR: https://search.r-project.org/CRAN/refmans/tuneR/html/readMidi.html
